cupy/sorting/search.py

```diff
@@ -213,29 +213,73 @@ def where(condition, x=None, y=None):
 
 
 _searchsorted_kernel = core.ElementwiseKernel(
-    'S x, raw T bins, int64 n_bins, bool side_is_right',
+    'S x, raw T bins, int64 n_bins, bool side_is_right, bool check_mono',
     'int64 y',
     '''
+    // Array is assumed to be monotonically
+    // increasing unless a check is requested
+    // because of functions like digitize
+    // allowing both, increasing and decreasing
+    bool inc = true;
+    if (check_mono && n_bins >= 2) {
```

Comment:
I think `inc` does not have to be calculated in each thread [as previously commented](https://github.com/cupy/cupy/pull/2726#discussion_r352017024).

Reply:
the original check copied from the histogram code took too long with large arrays as it compares all the elements.

One of the problems is that with the presence of nans and repeated values we can not solve it with comparing the first and the last elements and it becomes necessary to iterate the bins array.

[ 2, 2, 2, ... , 3, ..., nan, nan, ..., nan ]
is a valid monotonic increasing array
and it is not possible to check it by looking the first and the last element. 2<=nan is false. 

We can do this check outside the kernel, but is only a few lines and in most of the cases gets resolves with one comparison and the b values will be cached for later.

Let me know if you can think of a better idea 😃

Reply:
I think that you mean this from the other PR

>> I do think it would be helpful to have an option (by a thread-local config, not by a kwarg) to check monotonicity (with synchronization) and raise exception.

The problem with digitize is that it needs to know wether the array increases or decreases so the monotonicity check is always needed

Reply:
Hmm, I didn't mean checking only the first and the last, nor skipping the check.

Even if we are going to ensure monotonicity by checking all the elements, this check is redundant if done in each thread.

Reply:
I agree with it, but cost is low and all the threads in the warp executes the same code so I think is more efficient than firing another kernel. 99% of the cases it should be just only one comparison bins[0] <= bins[1]

comparing all the elements of the array and reducing them later had a pretty big impact in the execution time

Reply:
>99% of the cases it should be just only one comparison bins[0] <= bins[1]

Do you mean you assume `n_bins == 2` 99% of the time? Why?

Reply:
Sorry, I realized I didn't understand the loop correctly.

But anyway, my point remains the same.

>99% of the cases it should be just only one comparison bins[0] <= bins[1]

I'm not sure this estimation is right, but even so in the worst of the remaining 1% case, the performance will be prohibitively degraded. We can't assume that won't happen.
I don't think that's worth saving only a few kernel launches.

Reply:
it will be degraded anyway, because in such case the whole array should be iterated inside this kernel or not.

Reply:
>Maybe the first thread of the block can do it and store the result in shared memory for the rest of threads to read it?

That would require thread synchronization. It's essentially the same as the current implementation.

>Or i can just write another kernel for the monotonicity with the reduction phase optimized.

It doesn't have to be optimized.
You can just use something like `(bins[:-1] > bins[1:]).any()` and/or `(bins[:-1] < bins[1:]).any()`.
(Maybe special consideration about `nan` is needed. Perhaps `bins_ = bins.copy(); bins_[cp.isnan(bins_)] = np.inf`)

Reply:
I tried that before and the execution time went one order of magnitude up for large arrays.
(maybe there was some hidden synchronization I didn’t realize), thats why i tried the current approach.

Reply:
I tried that before and the execution time went one order of magnitude up for large arrays.
(maybe there was some hidden synchronization I didn’t realize), thats why i tried the current approach.

Reply:
Hmm, one order of magnitude of slowdown is not what I expected...
Let me think.

Reply:
Let me repeat again and show you the results, as usually maybe i just screwed it up somewhere 😥

Reply:
Let me repeat again and show you the results, as usually maybe i just screwed it up somewhere 😥

Reply:
I just pushed a change where I could get rid of the loop thinking just a bit of the nan comparison properties and how numpy defines the ordering.
It was possible to do it just by comparing 2 values 😅